python/trezorlib/eos.py

```diff
@@ -0,0 +1,353 @@
+        elif action['name'] == 'buyram':
+            txAction.buy_ram = parse_buy_ram(data)
+        elif action['name'] == 'buyrambytes':
+            txAction.buy_ram_bytes = parse_buy_rambytes(data)
+        elif action['name'] == 'sellram':
+            txAction.sell_ram = parse_sell_ram(data)
+        elif action['name'] == 'delegatebw':
+            txAction.delegate = parse_delegate(data)
+        elif action['name'] == 'undelegatebw':
+            txAction.undelegate = parse_undelegate(data)
+        elif action['name'] == 'refund':
+            txAction.refund = parse_refund(data)
+        elif action['name'] == 'updateauth':
+            txAction.update_auth = parse_updateauth(data)
+        elif action['name'] == 'deleteauth':
+            txAction.delete_auth = parse_deleteauth(data)
+        elif action['name'] == 'linkauth':
+            txAction.link_auth = parse_linkauth(data)
+        elif action['name'] == 'unlinkauth':
+            txAction.unlink_auth = parse_unlinkauth(data)
+        elif action['name'] == 'newaccount':
+            txAction.new_account = parse_new_account(data)
+    elif action['name'] == 'transfer':
+        txAction.transfer = parse_transfer(data)
+    else:
+        txAction.unknown = parse_unknown(data)
+
+    return txAction
+
+
+def parse_transaction_json(json):
+    tx = type('Transaction', (object,), {})()
+    tx.chain_id = bytes.fromhex(json['chain_id'])
+
+    body = json['transaction']
+  
+    expiration = int((datetime.strptime(body['expiration'], '%Y-%m-%dT%H:%M:%S') - datetime(1970, 1, 1)).total_seconds())
+    tx.expiration = expiration
+    tx.ref_block_num = int(body['ref_block_num'])
+    tx.ref_block_prefix = int(body['ref_block_prefix'])
+    tx.net_usage_words = int(body['net_usage_words'])
+    tx.max_cpu_usage_ms = int(body['max_cpu_usage_ms'])
+    tx.delay_sec = int(body['delay_sec'])
+
+    tx.actions = body['actions']
+
+    tx.num_actions = len(tx.actions)
+
+    return tx
+
+# ====== Client functions ====== #
+
+
+@expect(messages.EosPublicKey)
+def get_public_key(client, n, show_display=False, multisig=None):
+    response = client.call(messages.EosGetPublicKey(address_n=n, show_display=show_display))
+    return response
+
+
+@session
+def sign_tx(client, address, transaction):
+    tx = parse_transaction_json(transaction)
+
+    header = messages.EosTxHeader()
+    header.expiration = tx.expiration
+    header.ref_block_num = tx.ref_block_num
+    header.ref_block_prefix = tx.ref_block_prefix
+    header.max_net_usage_words = tx.net_usage_words
+    header.max_cpu_usage_ms = tx.max_cpu_usage_ms
+    header.delay_sec = tx.delay_sec
+
+    msg = messages.EosSignTx()
+    msg.address_n = address
+    msg.chain_id = tx.chain_id
+    msg.header = header
+    msg.num_actions = tx.num_actions
+
+    response = client.call(msg)
+
+    try:
+        while isinstance(response, messages.EosTxActionRequest):
+            response = client.call(parse_action(tx.actions.pop(0)))
```

Comment:
the `parse_action` call should happen in advance, otherwise you could crash in the middle of a flow
but see comment on `parse_transaction_json`, which would resolve this as well

Reply:
Restructured code. Done.